packages/services/api/src/modules/organization/providers/organization-access-tokens.ts

```diff
@@ -0,0 +1,363 @@
+      return {
+        type: 'error' as const,
+        message: 'Invalid input provided.',
+        details: {
+          title: titleResult.error?.issues.at(0)?.message ?? null,
+          description: descriptionResult.error?.issues.at(0)?.message ?? null,
+        },
+      };
+    }
+
+    const { organizationId } = await this.idTranslator.resolveOrganizationReference({
+      reference: args.organization,
+      onError() {
+        throw new InsufficientPermissionError('accessToken:modify');
+      },
+    });
+
+    await this.session.assertPerformAction({
+      organizationId,
+      params: { organizationId },
+      action: 'accessToken:modify',
+    });
+
+    const assignedResources =
+      await this.resourceAssignments.transformGraphQLResourceAssignmentInputToResourceAssignmentGroup(
+        organizationId,
+        args.assignedResources ?? { mode: 'granular' },
+      );
+
+    const permissions = Array.from(
+      new Set(
+        args.permissions.filter(permission => assignablePermissions.has(permission as Permission)),
+      ),
+    );
+
+    const id = crypto.randomUUID();
+    const accessKey = await OrganizationAccessKey.create(id);
+
+    const result = await this.pool.maybeOne<unknown>(sql`
+      INSERT INTO "organization_access_tokens" (
+        "id"
+        , "organization_id"
+        , "title"
+        , "description"
+        , "permissions"
+        , "assigned_resources"
+        , "hash"
+        , "first_characters"
+      )
+      VALUES (
+        ${id}
+        , ${organizationId}
+        , ${titleResult.data}
+        , ${descriptionResult.data}
+        , ${sql.array(permissions, 'text')}
+        , ${sql.jsonb(assignedResources)}
+        , ${accessKey.hash}
+        , ${accessKey.firstCharacters}
+      )
+      RETURNING
+        ${organizationAccessTokenFields}
+    `);
+
+    const organizationAccessToken = OrganizationAccessTokenModel.parse(result);
+
+    await this.cache.add(organizationAccessToken);
+
+    await this.auditLogs.record({
+      organizationId,
+      eventType: 'ORGANIZATION_ACCESS_TOKEN_CREATED',
+      metadata: {
+        organizationAccessTokenId: organizationAccessToken.id,
+      },
+    });
+
+    return {
+      type: 'success' as const,
+      organizationAccessToken,
+      privateAccessKey: accessKey.privateAccessToken,
+    };
+  }
+
+  async delete(args: { organizationAccessTokenId: string }) {
+    const record = await this.findById(args.organizationAccessTokenId);
+    if (record === null) {
+      throw new InsufficientPermissionError('accessToken:modify');
+    }
+
+    await this.session.assertPerformAction({
+      action: 'accessToken:modify',
+      organizationId: record.organizationId,
+      params: { organizationId: record.organizationId },
+    });
+
+    await this.pool.query(sql`
+      DELETE
+      FROM
+        "organization_access_tokens"
+      WHERE
+        "id" = ${args.organizationAccessTokenId}
+    `);
```

Comment:
Instead, we could also do a soft delete. We never did that AFAIK, what do you think?

Reply:
hahaha 😆 i was scrolling through the migration file and thought "should this be soft-deleted? feel like a sensitive info that might need to be kept" -> searched for this piece of code -> found your comment 😆

Reply:
And yeah, i do feel like we need to soft-delete

Reply:
@dotansimha What information should we store about the soft-delete?

We could store, the deletion date, user info (email; userId) etc. or we could just store a boolean `isDeleted`.

This information would be redundant with the audit log, but the audit log retention period is only a year.
The purpose of doing the soft delete would be to have more context if that one-year period has passed?

Reply:
If the goal is to save information for auditing or in case we need to find if a token was set for someone in the past, then I think it'd be better to move this to a separate storage mechanism.
I may be missing something, but I dont see a need to have this information so available. And soft deletes add complexity / are easy to mistakenly not check.

Reply:
I agree with the forgetting to do the check part when querying the database. 😅 @dotansimha we already store metadata within clickhouse, is the one-year retention insufficient?

Reply:
if we store it all in the audit log, and you think that there's no value in making it a soft-delete, then it's fine. <br>My main reason for writing that was because of audit log and trace for user actions. 

As long as users know it's permanently deleting it, we can skip soft-delete.